pkg/policy/api/egress.go

```diff
@@ -107,7 +107,25 @@ type EgressRule struct {
 	// + optional
 	ToServices []Service `json:"toServices,omitempty"`
 
-	// ToFQDN selects DNS targets.
+	// ToFQDN allows whitelisting DNS names in place of IPs. The IPs that result
+	// from DNS resolution of `ToFQDN.MatchName`s are added to the same
+	// EgressRule object as ToCIDRSet entries, and behave accordingly. Any L4 and
+	// L7 rules within this EgressRule will also apply to these IPs.
+	// The DNS -> IP mapping is re-resolved periodically from within the
+	// cilium-agent, and the IPs in the DNS response are effected in the policy
+	// for selected pods as-is (i.e. the list of IPs is not modified in any way).
+	// Note: An explicit rule to allow for DNS traffic is needed for the pods, as
+	// ToFQDN counts as an egress rule and will enforce egress policy when
+	// PolicyEnforcment=default.
+	//
+	// The current implementation has a number of limitations:
+	// - The DNS resolution originates from cilium-agent, and not from the pods.
+	// Differences between the responses seen by cilium agent and a particular
+	// pod will whitelist the incorrect IP.
+	// - DNS TTLs are ignored, and cilium-agent will repoll on a short interval
+	// (5 seconds). Each change to the DNS data will trigger a policy
+	// regeneration. This may result in delayed updates to the policy for an
+	// endpoint when the data changes often or the system is under load.
```

Comment:
I assume this TTL handling implies that there is no garbage collection as well. What's the implications of this?

I also wonder if there's also anything reasonable we can try to explain about how building your security policy against DNS resolution impacts the overall security of your system.

Reply:
In the history of this feature (and PR even) I've flip-flopped between accumulating IPs and treating each DNS lookup as the full list. Chatting with aanm, he reminded me that we would disconnect existing long-lived connections if we simply replaced one list of IPs with another when that list is different enough. `google.com` (a likely first test of this) does actually change every so often...

If we replace IPs as we do now, the rate of change of the IPs relative to connection lifetime is the biggest consideration.
If we accumulate all-seen IPs (the original plan based on my initial discussions) the risk is 1 incorrect IP would be retained forever (not to mention hitting our CIDR Identity limits). In most common cases, however, the list of IPs is relatively low.